File: backend/app/core/ticket_permissions.py

from dataclasses import dataclass
from enum import Enum
from typing import Dict, Set, Optional, FrozenSet

from app.models.ticket import TicketStatus


ROLE_SUPPORT_L1 = "SUPPORT_L1"
ROLE_SUPPORT_L2 = "SUPPORT_L2"
ROLE_SUPPORT_L3 = "SUPPORT_L3"
ROLE_ENGINEERING_MANAGER = "ENGINEERING_MANAGER"
ROLE_ADMIN = "ADMIN"


# ==========================================
# Permission Messages
# ==========================================
MSG_INVALID_ACTION = "Invalid or unknown action."
MSG_INVALID_TRANSITION = "Invalid lifecycle transition."
MSG_TICKET_CLOSED = "Ticket is already closed."
MSG_ROLE_REQUIRED = "You do not have the required role to perform this action."
MSG_ASSIGNEE_REQUIRED = "Only the assigned engineer may perform this action."
MSG_OVERRIDE_REQUIRED = "Administrator or Manager override required."


class TicketAction(str, Enum):
    ASSIGN = "ASSIGN"
    UNASSIGN = "UNASSIGN"
    START_PROGRESS = "START_PROGRESS"
    RESOLVE = "RESOLVE"
    CLOSE = "CLOSE"
    REOPEN = "REOPEN"
    CHANGE_PRIORITY = "CHANGE_PRIORITY"
    CHANGE_ASSIGNEE = "CHANGE_ASSIGNEE"

ALLOWED_TRANSITIONS: Dict[TicketStatus, FrozenSet[TicketStatus]] = {
    TicketStatus.OPEN: frozenset({TicketStatus.IN_PROGRESS}),
    TicketStatus.IN_PROGRESS: frozenset({TicketStatus.RESOLVED}),
    TicketStatus.RESOLVED: frozenset({TicketStatus.CLOSED}),
    TicketStatus.CLOSED: frozenset({TicketStatus.OPEN}),
}


@dataclass(frozen=True)
class PermissionMeta:
    allowed_roles: tuple[str, ...]
    requires_assignee: bool
    manager_override: bool
    admin_override: bool

ACTION_PERMISSIONS: Dict[TicketAction, PermissionMeta] = {
    TicketAction.ASSIGN: PermissionMeta(
        allowed_roles=(ROLE_ENGINEERING_MANAGER, ROLE_ADMIN),
        requires_assignee=False,
        manager_override=True,
        admin_override=True,
    ),
    TicketAction.UNASSIGN: PermissionMeta(
        allowed_roles=(ROLE_ENGINEERING_MANAGER, ROLE_ADMIN),
        requires_assignee=False,
        manager_override=True,
        admin_override=True,
    ),
    TicketAction.START_PROGRESS: PermissionMeta(
        allowed_roles=(ROLE_SUPPORT_L1, ROLE_SUPPORT_L2, ROLE_SUPPORT_L3),
        requires_assignee=True,
        manager_override=True,
        admin_override=True,
    ),
    TicketAction.RESOLVE: PermissionMeta(
        allowed_roles=(ROLE_SUPPORT_L1, ROLE_SUPPORT_L2, ROLE_SUPPORT_L3),
        requires_assignee=True,
        manager_override=True,
        admin_override=True,
    ),
    TicketAction.CLOSE: PermissionMeta(
        allowed_roles=(ROLE_ENGINEERING_MANAGER, ROLE_ADMIN),
        requires_assignee=False,
        manager_override=True,
        admin_override=True,
    ),
    TicketAction.REOPEN: PermissionMeta(
        allowed_roles=(ROLE_ENGINEERING_MANAGER, ROLE_ADMIN),
        requires_assignee=False,
        manager_override=True,
        admin_override=True,
    ),
    TicketAction.CHANGE_PRIORITY: PermissionMeta(
        allowed_roles=(ROLE_ENGINEERING_MANAGER, ROLE_ADMIN),
        requires_assignee=False,
        manager_override=True,
        admin_override=True,
    ),
    TicketAction.CHANGE_ASSIGNEE: PermissionMeta(
        allowed_roles=(ROLE_ENGINEERING_MANAGER, ROLE_ADMIN),
        requires_assignee=False,
        manager_override=True,
        admin_override=True,
    ),
}

TRANSITION_ACTIONS: Dict[tuple[TicketStatus, TicketStatus], TicketAction] = {
    (TicketStatus.OPEN, TicketStatus.IN_PROGRESS): TicketAction.START_PROGRESS,
    (TicketStatus.IN_PROGRESS, TicketStatus.RESOLVED): TicketAction.RESOLVE,
    (TicketStatus.RESOLVED, TicketStatus.CLOSED): TicketAction.CLOSE,
    (TicketStatus.CLOSED, TicketStatus.OPEN): TicketAction.REOPEN,
}
    

def is_transition_allowed(current_status: TicketStatus | str, new_status: TicketStatus | str) -> bool:
    """Check if transitioning from current_status to new_status is allowed in the matrix."""
    curr = TicketStatus(current_status) if isinstance(current_status, str) else current_status
    new_s = TicketStatus(new_status) if isinstance(new_status, str) else new_status
    return new_s in ALLOWED_TRANSITIONS.get(curr, set())

def get_allowed_transitions(current_status: TicketStatus | str) -> FrozenSet[TicketStatus]:
    """Get all allowed next statuses for a given status."""
    curr = TicketStatus(current_status) if isinstance(current_status, str) else current_status
    return ALLOWED_TRANSITIONS.get(curr, frozenset())

def get_action_for_transition(current_status: TicketStatus | str, new_status: TicketStatus | str) -> Optional[TicketAction]:
    """Get the action representing a specific status transition, if valid."""
    curr = TicketStatus(current_status) if isinstance(current_status, str) else current_status
    new_s = TicketStatus(new_status) if isinstance(new_status, str) else new_status
    return TRANSITION_ACTIONS.get((curr, new_s))

def get_permission_meta(action: TicketAction | str) -> Optional[PermissionMeta]:
    """Get the permission metadata for a given action."""
    act = TicketAction(action) if isinstance(action, str) else action
    return ACTION_PERMISSIONS.get(act)

def action_requires_assignee(action: TicketAction | str) -> bool:
    """Check if the given action requires the user to be the current assignee of the ticket."""
    meta = get_permission_meta(action)
    return meta.requires_assignee if meta else False

def manager_override_allowed(action: TicketAction | str) -> bool:
    """Check if a manager can override the standard rules for this action."""
    meta = get_permission_meta(action)
    return meta.manager_override if meta else False

def admin_override_allowed(action: TicketAction | str) -> bool:
    """Check if an admin can override the standard rules for this action."""
    meta = get_permission_meta(action)
    return meta.admin_override if meta else False

def get_allowed_roles_for_action(action: TicketAction | str) -> tuple[str, ...]:
    """Get the standard allowed roles for a given action."""
    meta = get_permission_meta(action)
    return meta.allowed_roles if meta else ()

File: frontend/src/lib/ticketPermissions.ts
import type { CurrentUser } from "../types/auth";
import { Roles } from "../types/auth";
import type { TicketRead, TicketStatus } from "../types/ticket";
import type { TicketPermissions } from "../types/permissions";

type TicketAction = 
  | "ASSIGN"
  | "UNASSIGN"
  | "START_PROGRESS"
  | "RESOLVE"
  | "CLOSE"
  | "REOPEN"
  | "CHANGE_PRIORITY"
  | "CHANGE_ASSIGNEE";

interface PermissionMeta {
  allowedRoles: string[];
  requiresAssignee: boolean;
  managerOverride: boolean;
  adminOverride: boolean;
}

const ACTION_PERMISSIONS: Record<TicketAction, PermissionMeta> = {
  ASSIGN: {
    allowedRoles: [Roles.ENGINEERING_MANAGER, Roles.ADMIN],
    requiresAssignee: false,
    managerOverride: true,
    adminOverride: true,
  },
  UNASSIGN: {
    allowedRoles: [Roles.ENGINEERING_MANAGER, Roles.ADMIN],
    requiresAssignee: false,
    managerOverride: true,
    adminOverride: true,
  },
  START_PROGRESS: {
    allowedRoles: [Roles.SUPPORT_L1, Roles.SUPPORT_L2, Roles.SUPPORT_L3],
    requiresAssignee: true,
    managerOverride: true,
    adminOverride: true,
  },
  RESOLVE: {
    allowedRoles: [Roles.SUPPORT_L1, Roles.SUPPORT_L2, Roles.SUPPORT_L3],
    requiresAssignee: true,
    managerOverride: true,
    adminOverride: true,
  },
  CLOSE: {
    allowedRoles: [Roles.ENGINEERING_MANAGER, Roles.ADMIN],
    requiresAssignee: false,
    managerOverride: true,
    adminOverride: true,
  },
  REOPEN: {
    allowedRoles: [Roles.ENGINEERING_MANAGER, Roles.ADMIN],
    requiresAssignee: false,
    managerOverride: true,
    adminOverride: true,
  },
  CHANGE_PRIORITY: {
    allowedRoles: [Roles.ENGINEERING_MANAGER, Roles.ADMIN],
    requiresAssignee: false,
    managerOverride: true,
    adminOverride: true,
  },
  CHANGE_ASSIGNEE: {
    allowedRoles: [Roles.ENGINEERING_MANAGER, Roles.ADMIN],
    requiresAssignee: false,
    managerOverride: true,
    adminOverride: true,
  },
};

const ALLOWED_TRANSITIONS: Record<TicketStatus, TicketStatus[]> = {
  OPEN: ["IN_PROGRESS"],
  IN_PROGRESS: ["RESOLVED"],
  RESOLVED: ["CLOSED"],
  CLOSED: ["OPEN"],
};

const getUserRoles = (user: CurrentUser | null): Set<string> => {
  if (!user || !user.roles) return new Set();
  return new Set(
    user.roles.map((r) => typeof r === 'string' ? r.toUpperCase() : (r as any).name?.toUpperCase())
  );
};

const ROLE_RANK: Record<string, number> = {
  [Roles.SUPPORT_L1]: 1,
  [Roles.SUPPORT_L2]: 2,
  [Roles.SUPPORT_L3]: 3,
  [Roles.ENGINEERING_MANAGER]: 4,
  [Roles.ADMIN]: 5,
};

const getUserHighestRank = (userRoles: Set<string>): number => {
  let maxRank = 0;
  userRoles.forEach(role => {
    const rank = ROLE_RANK[role] || 0;
    if (rank > maxRank) maxRank = rank;
  });
  return maxRank;
};

const checkTransition = (ticket: TicketRead, newStatus?: TicketStatus): boolean => {
  if (newStatus) {
    if (newStatus === ticket.status) return false;
    return (ALLOWED_TRANSITIONS[ticket.status] || []).includes(newStatus);
  }
  return true;
};

const checkOverride = (userRoles: Set<string>, meta?: PermissionMeta): boolean => {
  if (meta) {
    if (userRoles.has(Roles.ADMIN) && meta.adminOverride) return true;
    if (userRoles.has(Roles.ENGINEERING_MANAGER) && meta.managerOverride) return true;
    return false;
  }
  // No meta: unconditionally grant override to Admin and Manager
  return userRoles.has(Roles.ADMIN) || userRoles.has(Roles.ENGINEERING_MANAGER);
};

const checkRoles = (userRoles: Set<string>, meta: PermissionMeta): boolean => {
  return meta.allowedRoles.some((role) => userRoles.has(role));
};

const checkAssignee = (ticket: TicketRead, user: CurrentUser | null, meta: PermissionMeta): boolean => {
  if (meta.requiresAssignee) {
    if (!user || ticket.assigned_to_id !== user.id) {
      return false;
    }
  }
  return true;
};

const validateAction = (
  user: CurrentUser | null,
  ticket: TicketRead,
  action: TicketAction,
  newStatus?: TicketStatus
): boolean => {
  const meta = ACTION_PERMISSIONS[action];
  if (!meta) return false;

  if (!checkTransition(ticket, newStatus)) return false;

  const userRoles = getUserRoles(user);

  if (checkOverride(userRoles, meta)) return true;
  if (!checkRoles(userRoles, meta)) return false;
  if (!checkAssignee(ticket, user, meta)) return false;

  return true;
};

/**
 * Escalation is not a lifecycle transition — it promotes the ticket's support level.
 * Mirrors backend domain/permissions.py::can_escalate():
 *   - L3 tickets: never escalatable (no next level exists)
 *   - L1 tickets: requires rank >= 1 (any support engineer or above)
 *   - L2 tickets: requires rank >= 2 (L2, L3, Manager, Admin)
 *   - Admin/Manager override: allowed at any valid level
 * NOTE: The backend does NOT require the actor to be the assignee for escalation.
 */
const validateEscalate = (user: CurrentUser | null, ticket: TicketRead): boolean => {
  // 1. Business constraint: L3 has no next level — hard stop before any role check
  if (ticket.support_level === "L3") return false;

  const userRoles = getUserRoles(user);

  // 2. Admin/Manager override (mirrors backend rank >= 4/5)
  if (checkOverride(userRoles)) return true;

  // 3. Rank-based check — mirrors backend can_escalate()
  const userRank = getUserHighestRank(userRoles);
  const requiredRank = ticket.support_level === "L1" ? 1 : ticket.support_level === "L2" ? 2 : 99;

  return userRank >= requiredRank;
};

export const getTicketPermissions = (user: CurrentUser | null, ticket: TicketRead): TicketPermissions => {
  const canAssign = validateAction(user, ticket, "ASSIGN");
  const canChangeAssignee = validateAction(user, ticket, "CHANGE_ASSIGNEE");
  
  const canStartProgress = validateAction(user, ticket, "START_PROGRESS", "IN_PROGRESS");
  const canResolve = validateAction(user, ticket, "RESOLVE", "RESOLVED");
  const canClose = validateAction(user, ticket, "CLOSE", "CLOSED");
  const canReopen = validateAction(user, ticket, "REOPEN", "OPEN");

  const canEditPriority = validateAction(user, ticket, "CHANGE_PRIORITY");
  const canEscalate = validateEscalate(user, ticket);

  return {
    canAssign,
    canChangeAssignee,
    canStartProgress,
    canResolve,
    canClose,
    canReopen,
    canEditPriority,
    canEscalate,
  };
};
